### PR TITLE
Implements TileMap layers and move TileSetPlugins's functions to the TileMap node instead

### PR DESCRIPTION
**Pull request by groud**
This implements TileMap layers. Layers are now part of the TileMap node. Here is the list of their features:
- Layers can be named.
- Enabling/disabling layers. This is mainly to make some layers visible.
- A layer can set set to Y-sort its tiles. In such case, you can specify a Y-sort origin integer value that will be added to the tiles Y-sort origin. This allow using layers for different height values, which is quite useful for top-down games.
- A layer can have a Z-index value that will be added to the tiles Z-index. This allow some layers to be drawn on top/bottom of others. This is especially useful when some layers are Y-sorted.
- The editors features a button to highlight the currently selected layer. This makes layers belong the selected one darker and the one on top of it translucent.

As a side note, I also removed the TileSetPlugin code and moved that to the TileMap node instead. This makes things easier to implement while still keeping different functions for different system, which keeps things organized.

Here it is in action:
Peek 29-07-2021 11-45

### CODE REVIEW

File: editor/plugins/tiles/tile_map_editor.cpp [hunk 1]
```diff
@@ -3516,33 +3725,60 @@ TileMapEditor::TileMapEditor() {
...
+	layers_selection_button = memnew(Button);
+	layers_selection_button->set_toggle_mode(true);
+	layers_selection_button->connect("draw", callable_mp(this, &TileMapEditor::_layers_selection_button_draw));
+	layers_selection_button->connect("pressed", callable_mp(this, &TileMapEditor::_layers_selection_button_pressed));
+	layers_selection_button->connect("hidden", callable_mp((Window *)layers_selection_popup, &Popup::hide));
+	layers_selection_button->set_tooltip(TTR("Tile Map Layer"));
+	layers_selection_button->add_child(layers_selection_popup);
+	tile_map_toolbar->add_child(layers_selection_button);
+
+	toogle_highlight_selected_layer_button = memnew(Button);
```
Comment (YeldhamDev) [inline editor/plugins/tiles/tile_map_editor.cpp:3759]: This button could use a tooltip as well.

**Comment by YeldhamDev** [comment]
Don' forget to minimize the SVG files.

**Comment by KoBeWi** [comment]
I gave it a quick test and it's great! The only missing thing would be some shortcut to quickly switch between layers (if there isn't one already). Something like Ctrl + 1-9 maybe. I spotted a few issues though:

Not sure if it's intended, but the layer highlight is still active when you deselect the TileMap. Shouldn't it get "disabled" to display the TileMap normally if it's not selected?

The binding for `set_cell()` wasn't updated and it doesn't include the new `layer` argument.

Uh, maybe it's unrelated or non-issue, but feels to me like the dropdown has too much free space on the left:
image
idk

**Comment by groud** [comment]
> I gave it a quick test and it's great! The only missing thing would be some shortcut to quickly switch between layers (if there isn't one already). Something like Ctrl + 1-9 maybe. 

I implemented layer switching with page_up, page_down. I think it's enough.

> Not sure if it's intended, but the layer highlight is still active when you deselect the TileMap. Shouldn't it get "disabled" to display the TileMap normally if it's not selected?

Solved that one, layer selection is only visible when the TileMap node is active now.

> Uh, maybe it's unrelated or non-issue, but feels to me like the dropdown has too much free space on the left

Well, I compute the width of the button from the minimal width of the several layer names. This avoid the button to change size depending on the inside value. There is some margin too but I can't do much about it without overriding theme elements (which bothers me). So well, I kept it that way.

**Review by KoBeWi — APPROVED** [review]
Looks fine functionality-wise. I didn't check the code though.

File: doc/classes/TileMap.xml [hunk 2]
```diff
@@ -65,8 +86,8 @@
 		</method>
 		<method name="get_used_cells" qualifiers="const">
 			<return type="Vector2i[]" />
+			<argument index="0" name="layer" type="int" />
 			<description>
-				Returns a [Vector2] array with the positions of all cells containing a tile from the tileset (i.e. a tile index different from [code]-1[/code]).
```
Comment (akien-mga) [inline doc/classes/TileMap.xml:69]: You lost documentation here.

File: doc/classes/TileMap.xml [hunk 3]
```diff
@@ -84,18 +117,47 @@
...
+			<argument index="1" name="enabled" type="bool" />
+			<description>
+			</description>
+		</method>
+		<method name="set_layer_name">
+			<return type="void" />
+			<argument index="0" name="layer" type="int" />
+			<argument index="1" name="name" type="String" />
 			<description>
-				Sets the tile index for the cell given by a Vector2i.
```
Comment (akien-mga) [inline doc/classes/TileMap.xml:92]: Lost docs.

File: doc/classes/TileMap.xml [hunk 4]
```diff
@@ -84,18 +117,47 @@
...
+			<argument index="1" name="y_sort_origin" type="int" />
+			<description>
+			</description>
+		</method>
+		<method name="set_layer_z_index">
+			<return type="void" />
+			<argument index="0" name="layer" type="int" />
+			<argument index="1" name="z_index" type="int" />
 			<description>
-				Updates the tile map's quadrants, allowing things such as navigation and collision shapes to be immediately used if modified.
```
Comment (akien-mga) [inline doc/classes/TileMap.xml:98]: Lost docs.

Comment (groud) [reply doc/classes/TileMap.xml:98]: I removed this method. It's was supposed to be internal I guess.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**